src/main/java/ee/carlrobert/llm/client/watsonx/WatsonxClient.java

```diff
@@ -0,0 +1,163 @@
+import ee.carlrobert.llm.PropertiesLoader;
+import ee.carlrobert.llm.client.DeserializationUtil;
+import ee.carlrobert.llm.client.openai.completion.ErrorDetails;
+import ee.carlrobert.llm.client.watsonx.completion.WatsonxCompletionRequest;
+import ee.carlrobert.llm.client.watsonx.completion.WatsonxCompletionResponse;
+import ee.carlrobert.llm.client.watsonx.completion.WatsonxCompletionResponseError;
+import ee.carlrobert.llm.completion.CompletionEventListener;
+import ee.carlrobert.llm.completion.CompletionEventSourceListener;
+import okhttp3.*;
+import okhttp3.sse.EventSource;
+import okhttp3.sse.EventSources;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static ee.carlrobert.llm.client.DeserializationUtil.OBJECT_MAPPER;
+
+public class WatsonxClient {
+
+        private static final MediaType APPLICATION_JSON = MediaType.parse("application/json");
+        private final OkHttpClient httpClient;
+        private final String host;
+        private final String apiVersion;
+        private final WatsonxAuthenticator authenticator;
+
+        private WatsonxClient(Builder builder, OkHttpClient.Builder httpClientBuilder) {
+            this.httpClient = httpClientBuilder.build();
+            this.apiVersion = builder.apiVersion;
+            this.host = builder.host;
+            if (builder.isOnPrem) {
+                if (builder.isZenApiKey)
+                    this.authenticator = new WatsonxAuthenticator(builder.username, builder.apiKey);
+                else
+                    this.authenticator = new WatsonxAuthenticator(builder.username, builder.apiKey, builder.host);
+            } else {
+                this.authenticator = new WatsonxAuthenticator(builder.apiKey);
+            }
+        }
+
+       public EventSource getCompletionAsync(
+            WatsonxCompletionRequest request,
+            CompletionEventListener<String> eventListener) {
+        return EventSources.createFactory(httpClient).newEventSource(
+                buildCompletionRequest(request),
+                getCompletionEventSourceListener(eventListener));
+    }
+
+    public WatsonxCompletionResponse getCompletion(WatsonxCompletionRequest request) {
+        try (var response = httpClient.newCall(buildCompletionRequest(request)).execute()) {
+            return DeserializationUtil.mapResponse(response, WatsonxCompletionResponse.class);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected Request buildCompletionRequest(WatsonxCompletionRequest request) {
+        var headers = new HashMap<>(getRequiredHeaders());
+        if (request.getStream()) {
+            headers.put("Accept", "text/event-stream");
+        }
+        try {
+            return new Request.Builder()
+                    .url(host + "/ml/v1/text/" + (request.getStream() ? "generation_stream" : "generation") + "?version=" + apiVersion)
```

Comment:
imo, multi ternaries are ugly

Reply:
it's a single ternary that looks like an ugly multi ternary :)